stl/inc/ranges

```diff
@@ -22,15 +25,462 @@ _STL_DISABLE_CLANG_WARNINGS
+        _Semiregular_box& operator=(_Semiregular_box&&) requires movable<_Ty> = default;
+
+        _Semiregular_box& operator=(const _Semiregular_box& _That) noexcept {
+            _Val.~_Ty();
+            _Construct_in_place(_Val, _That._Val);
+            return *this;
+        }
+
+        _Semiregular_box& operator=(_Semiregular_box&& _That) noexcept {
+            _Val.~_Ty();
+            _Construct_in_place(_Val, _STD move(_That._Val));
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return true;
+        }
+
+        _NODISCARD constexpr _Ty& operator*() noexcept {
+            return _Val;
+        }
+        _NODISCARD constexpr const _Ty& operator*() const noexcept {
+            return _Val;
+        }
+
+    private:
+        /* [[no_unique_address]] */ _Ty _Val = _Ty();
+    };
+#else // ^^^ no workaround / workaround vvv
+    template <class _Ty>
+    class _Semiregular_box_copy : public optional<_Ty> {
+    public:
+        using optional<_Ty>::optional;
+
+        _Semiregular_box_copy()                             = default;
+        _Semiregular_box_copy(const _Semiregular_box_copy&) = default;
+        _Semiregular_box_copy(_Semiregular_box_copy&&)      = default;
+        _Semiregular_box_copy& operator=(_Semiregular_box_copy&&) = default;
+
+        _Semiregular_box_copy& operator=(const _Semiregular_box_copy& _That) noexcept(
+            is_nothrow_copy_constructible_v<_Ty>) {
+            optional<_Ty>::reset();
+            if (_That) {
+                optional<_Ty>::emplace(*_That);
+            }
+            return *this;
+        }
+    };
+
+    template <class _Ty>
+    using _Choose_semiregular_box_copy =
+        conditional_t<assignable_from<_Ty&, const _Ty&>, optional<_Ty>, _Semiregular_box_copy<_Ty>>;
+
+    template <class _Ty>
+    class _Semiregular_box_move : public _Choose_semiregular_box_copy<_Ty> {
+    public:
+        using _Choose_semiregular_box_copy<_Ty>::_Choose_semiregular_box_copy;
+
+        _Semiregular_box_move()                             = default;
+        _Semiregular_box_move(const _Semiregular_box_move&) = default;
+        _Semiregular_box_move(_Semiregular_box_move&&)      = default;
+        _Semiregular_box_move& operator=(const _Semiregular_box_move&) = default;
+
+        _Semiregular_box_move& operator=(_Semiregular_box_move&& _That) noexcept(is_nothrow_move_constructible_v<_Ty>) {
+            optional<_Ty>::reset();
+            if (_That) {
+                optional<_Ty>::emplace(_STD move(*_That));
+            }
+            return *this;
+        }
+    };
+
+    template <class _Ty>
+    using _Choose_semiregular_box_move =
+        conditional_t<assignable_from<_Ty&, _Ty>, _Choose_semiregular_box_copy<_Ty>, _Semiregular_box_move<_Ty>>;
+
+    template <_Copy_constructible_object _Ty>
+    class _Semiregular_box : public _Choose_semiregular_box_move<_Ty> {

```

Comment:
This class seems like it could really benefit from `_Mybase`.

Reply:
Ditto linewrapping and preference to minimize obfuscation, but I'll agree that `_Choose_semiregular_box_move<_Ty>` is a mouthful. If there were a fourth repetition I probably would define an alias here to drop from four to two, but eliminating a single use of an awkward name hardly seems worth adding two more lines.

This isn't a hill worth dying for, however - I'll cave if it's important to you.